### PR TITLE
Hardfork countdown support

### PR DESCRIPTION
**Pull request by paweljakubas**
# Issue Number

#1917 

# Overview

- [x] I have extended `ApiNetworkParameters` 
- [x] I have updated swagger accordingly
- [x] I have updated core `TypeSpec`
- [x] I have extended `NetworkParameters`
- [x] I have extended DB schema and addressed DB migrations
- [x] I have used `QueryAnytimeShelley EraStart` from `CardanoQuery` to get value
- [x] I have tested the above

# Comments

To move forward I need https://github.com/input-output-hk/cardano-wallet/pull/1909 to be merged. 
As,
```
QueryAnyTimeShelley EraStart :: Query CardanoBlock
```

### CODE REVIEW

File: lib/core/src/Cardano/Wallet/Api/Types.hs [hunk 1]
```diff
@@ -530,6 +545,17 @@ toApiNetworkParameters (NetworkParameters gp pp) = ApiNetworkParameters
...
     (view #desiredNumberOfStakePools pp)
     (Quantity $ fromIntegral $ getCoin $ view #minimumUTxOvalue pp)
+    (fromEraTransition <$> (view #transitionEras pp))
+  where
+    fromEraTransition (EraTransition info (Just epochN)) =
+        ApiEraTransition
+        (ApiEraTransitionInfo $ ApiT info)
+        (Just $ ApiEpochInfo
+            (ApiT epochN)
+            (Slotting.epochStartTime (Slotting.slotParams gp) epochN)
```
Comment (Anviking) [inline lib/core/src/Cardano/Wallet/Api/Types.hs]: Be aware of this: https://github.com/input-output-hk/cardano-wallet/pull/1913#discussion_r456778755

**Comment by paweljakubas** [comment]
bors try

File: lib/core/src/Cardano/Wallet/Api/Types.hs [hunk 2]
```diff
@@ -504,6 +509,15 @@ data ApiFee = ApiFee
     , estimatedMax :: !(Quantity "lovelace" Natural)
     } deriving (Eq, Generic, Show)
 
+newtype ApiEraTransitionInfo = ApiEraTransitionInfo
+    { transitionInfo :: (ApiT EraTransitionInfo)
+    } deriving (Eq, Generic, Show)
+
+data ApiEraTransition = ApiEraTransition
+    { transition :: !ApiEraTransitionInfo
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Types.hs]: Why not simply "ApiT EraTransitionInfo" here? What does the indirection gives us?

**Comment by paweljakubas** [comment]
bors try

File: lib/core/src/Cardano/Wallet/Api/Types.hs [hunk 3]
```diff
@@ -530,6 +532,12 @@ toApiNetworkParameters (NetworkParameters gp pp) = ApiNetworkParameters
     (Quantity $ unDecentralizationLevel $ view #decentralizationLevel pp)
     (view #desiredNumberOfStakePools pp)
     (Quantity $ fromIntegral $ getCoin $ view #minimumUTxOvalue pp)
+    (toApiEpochNo <$> view #hardforkEpochNo pp)
+  where
+    toApiEpochNo epochN =
+        ApiEpochInfo
+        (ApiT epochN)
+        (Slotting.epochStartTime (Slotting.slotParams gp) epochN)
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Types.hs]: This needs to use the time interpreter.

File: lib/core/src/Cardano/Wallet/DB/Sqlite.hs [hunk 4]
```diff
@@ -458,6 +460,17 @@ migrateManually tr defaultFieldValues =
...
 
+    -- | Adds an 'hardfork_epoch' column to the 'protocol_parameters'
+    -- table if it is missing.
+    --
+    addHardforkEpochIfMissing :: Sqlite.Connection -> IO ()
+    addHardforkEpochIfMissing conn = do
+        addColumn conn (DBField ProtocolParametersHardforkEpoch) value
+      where
+        value = case defaultHardforkEpoch defaultFieldValues of
+            Nothing -> "NULL"
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/DB/Sqlite.hs:471]: Does this work fine? Have you run the automated migration tests?

File: lib/core/src/Cardano/Wallet/DB/Sqlite/Types.hs [hunk 5]
```diff
@@ -288,17 +289,41 @@ instance PersistField SlotNo where
...
+    | isValidEpochNo c = Right c
+    | otherwise = Left . T.pack $ "not a valid epoch number: " <> show n
+    where c = unsafeEpochNo n
+
+instance PersistField EpochNo where
+    toPersistValue = toPersistValue . fromIntegral @Word31 @Word32 . unEpochNo
+    fromPersistValue = fromPersistValue >=> mkEpochNo
+
+instance PersistFieldSql EpochNo where
+    sqlType _ = sqlType (Proxy @Text)
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/DB/Sqlite/Types.hs]: :thinking: Why is the epoch number stored as a `Text` ? Wouldn't something like a Word32 or Word64 be more suitable (especially for things like filtering and where clauses)?

File: lib/shelley/src/Cardano/Wallet/Shelley/Network.hs [hunk 6]
```diff
@@ -645,21 +645,28 @@ mkTipSyncClient tr np localTxSubmissionQ onTipUpdate onPParamsUpdate onInterpret
...
+            Left (e1 :: AcquireFailure) ->
+                traceWith tr $ MsgLocalStateQueryError TipSyncClient $ show e1
+            Right (Right ls) -> do
+                    boundRes <- localStateQueryQ `send`
+                        CmdQueryLocalState pt (QueryAnytimeShelley GetEraStart)
+                    case boundRes of
+                        Right boundM ->
+                            onPParamsUpdate' $ convert boundM ls
+                        Left (e2 :: AcquireFailure) ->
+                            traceWith tr $ MsgLocalStateQueryError TipSyncClient $ show e2
```
Comment (KtorZ) [inline lib/shelley/src/Cardano/Wallet/Shelley/Network.hs]: Hmmm. The logic feels sort of weird / intertwined here. Query data is mixed with handling results which makes the whole a bit hard to follow (+ causes some logic duplication in the error handling). 

What about:

- run the GetCurrentPParams & GetEraStart in "parallel"
- combine them via an applicative together in `handleParamsUpdate`

So something along the lines of:

```hs
queryLocalState
    :: Point (CardanoBlock sc)
    -> m ()
queryLocalState pt = do
    mb <- localStateQueryQ `send`
        CmdQueryLocalState pt (QueryAnytimeShelley GetEraStart)

    pp <- localStateQueryQ `send`
        CmdQueryLocalState pt (QueryIfCurrentShelley Shelley.GetCurrentPParams)

    sequence (handleParamsUpdate fromShelleyPParams <$> mb <*> pp)
        >>= handleAcquireFailure

    st <- localStateQueryQ `send`
        CmdQueryLocalState pt (QueryIfCurrentByron Byron.GetUpdateInterfaceState)

    sequence (handleParamsUpdate protocolParametersFromUpdateState <$> mb <*> st)
        >>= handleAcquireFailure

handleAcquireFailure
    :: Either AcquireFailure ()
    -> m ()
handleAcquireFailure = \case
    Right () ->
        pure ()
    Left e ->
        traceWith tr $ MsgLocalStateQueryError TipSyncClient $ show e

handleParamsUpdate
    :: (Maybe Bound -> p -> W.ProtocolParameters)
    -> (Maybe Bound)
    -> (Either (MismatchEraInfo (CardanoEras sc)) p)
    -> m ()
handleParamsUpdate convert boundM = \case
    Right ls -> do
        onPParamsUpdate' $ convert boundM ls
    Left mismatch -> do
        traceWith tr $ MsgLocalStateQueryEraMismatch mismatch 
```

File: lib/core/src/Cardano/Wallet/DB/Sqlite/Types.hs (hunk 5 above)
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/DB/Sqlite/Types.hs]: frankly speaking I was here mimicking how we handled this for `SlotNo`

File: lib/shelley/src/Cardano/Wallet/Shelley/Network.hs (hunk 6 above)
Comment (paweljakubas) [reply lib/shelley/src/Cardano/Wallet/Shelley/Network.hs]: yes, it is much better. I was planning to do it after proof of concept solution

File: lib/core/src/Cardano/Wallet/Api/Types.hs (hunk 2 above)
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/Api/Types.hs]: this was removed on saturday, I do not know how you saw it as I squashed all commits before.

File: lib/shelley/src/Cardano/Wallet/Shelley/Network.hs (hunk 6 above)
Comment (paweljakubas) [reply lib/shelley/src/Cardano/Wallet/Shelley/Network.hs]: done in https://github.com/input-output-hk/cardano-wallet/pull/1932/commits/611ac058eab27fc8eb60898f47bda56bfbb6aba0

File: lib/core/src/Cardano/Wallet/Api/Types.hs (hunk 3 above)
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/Api/Types.hs]: done in https://github.com/input-output-hk/cardano-wallet/pull/1932/commits/35f63853178f1b39e9c98f85b787b09d1b15625e

File: lib/core/src/Cardano/Wallet/DB/Sqlite/Types.hs (hunk 5 above)
Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/DB/Sqlite/Types.hs]: We do store `SlotNo` as `Word64` though :sweat_smile: 

```hs
deriving via Word64 instance (PersistFieldSql SlotNo)
```

Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/DB/Sqlite/Types.hs]: yes, very true -> done in https://github.com/input-output-hk/cardano-wallet/pull/1932/commits/a3a5f8fb20de51f5dbf4389b7c462c59435f1d73

File: lib/core/src/Cardano/Wallet/DB/Sqlite.hs (hunk 4 above)
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/DB/Sqlite.hs:471]: so I run : 
```
nix-build nix/migration-tests.nix -o migration-tests
./migration-tests/runall.sh
```
and I see some failures: 
```
[cardano-wallet.pools-db:Error:13] [2020-07-27 14:01:13.46 UTC] Failed to migrate the database: PersistError "\n\nDatabase migration: manual intervention required.\nThe unsafe actions are prefixed by '***' below:\n\n    CREATE TEMP TABLE \"pool_owner_backup\"(\"pool_id\" VARCHAR NOT NULL,\"slot\" INTEGER NOT NULL,\"slot_internal_index\" INTEGER NOT NULL,\"pool_owner\" VARCHAR NOT NULL,\"pool_owner_index\" INTEGER NOT NULL, PRIMARY KEY (\"pool_id\",\"slot\",\"slot_internal_index\",\"pool_owner\",\"pool_owner_index\"));\n    INSERT INTO \"pool_owner_backup\"(\"pool_id\",\"pool_owner\",\"pool_owner_index\") SELECT \"pool_id\",\"pool_owner\",\"pool_owner_index\" FROM \"pool_owner\";\n    DROP TABLE \"pool_owner\";\n    CREATE TABLE \"pool_owner\"(\"pool_id\" VARCHAR NOT NULL,\"slot\" INTEGER NOT NULL,\"slot_internal_index\" INTEGER NOT NULL,\"pool_owner\" VARCHAR NOT NULL,\"pool_owner_index\" INTEGER NOT NULL, PRIMARY KEY (\"pool_id\",\"slot\",\"slot_internal_index\",\"pool_owner\",\"pool_owner_index\"), CONSTRAINT \"pool_ownerfk_registration_pool_id\" FOREIGN KEY(\"pool_id\",\"slot\",\"slot_internal_index\") REFERENCES \"pool_registration\"(\"pool_id\",\"slot\",\"slot_internal_index\") ON DELETE CASCADE);\n    INSERT INTO \"pool_owner\" SELECT \"pool_id\",\"slot\",\"slot_internal_index\",\"pool_owner\",\"pool_owner_index\" FROM \"pool_owner_backup\";\n    DROP TABLE \"pool_owner_backup\";\n    CREATE TEMP TABLE \"pool_registration_backup\"(\"pool_id\" VARCHAR NOT NULL,\"slot\" INTEGER NOT NULL,\"slot_internal_index\" INTEGER NOT NULL,\"margin_numerator\" INTEGER NOT NULL,\"margin_denominator\" INTEGER NOT NULL,\"cost\" INTEGER NOT NULL,\"pledge\" INTEGER NOT NULL,\"metadata_url\" VARCHAR NULL,\"metadata_hash\" VARCHAR NULL, PRIMARY KEY (\"pool_id\",\"slot\",\"slot_internal_index\"));\n    INSERT INTO \"pool_registration_backup\"(\"pool_id\",\"slot\",\"cost\") SELECT \"pool_id\",\"slot\",\"cost\" FROM \"pool_registration\";\n*** DROP TABLE \"pool_registration\";\n    CREATE TABLE \"pool_registration\"(\"pool_id\" VARCHAR NOT NULL,\"slot\" INTEGER NOT NULL,\"slot_internal_index\" INTEGER NOT NULL,\"margin_numerator\" INTEGER NOT NULL,\"margin_denominator\" INTEGER NOT NULL,\"cost\" INTEGER NOT NULL,\"pledge\" INTEGER NOT NULL,\"metadata_url\" VARCHAR NULL,\"metadata_hash\" VARCHAR NULL, PRIMARY KEY (\"pool_id\",\"slot\",\"slot_internal_index\"));\n    INSERT INTO \"pool_registration\" SELECT \"pool_id\",\"slot\",\"slot_internal_index\",\"margin_numerator\",\"margin_denominator\",\"cost\",\"pledge\",\"metadata_url\",\"metadata_hash\" FROM \"pool_registration_backup\";\n    DROP TABLE \"pool_registration_backup\";\n    CREATE TABLE \"pool_retirement\"(\"pool_id\" VARCHAR NOT NULL,\"slot\" INTEGER NOT NULL,\"slot_internal_index\" INTEGER NOT NULL,\"epoch\" INTEGER NOT NULL, PRIMARY KEY (\"pool_id\",\"slot\",\"slot_internal_index\"));\n    CREATE TABLE \"pool_metadata\"(\"metadata_hash\" VARCHAR NOT NULL,\"name\" VARCHAR NOT NULL,\"ticker\" VARCHAR NOT NULL,\"description\" VARCHAR NULL,\"homepage\" VARCHAR NOT NULL, PRIMARY KEY (\"metadata_hash\"));\n    CREATE TABLE \"pool_metadata_fetch_attempts\"(\"metadata_hash\" VARCHAR NOT NULL,\"metadata_url\" VARCHAR NOT NULL,\"retry_after\" TIMESTAMP NOT NULL,\"retry_count\" INTEGER NOT NULL, PRIMARY KEY (\"metadata_hash\",\"metadata_url\"));\n"
```
also
```
[cardano-wallet.pools-db:Error:13] [2020-07-27 14:00:50.77 UTC] Failed to migrate the database: : NOT NULL constraint failed: pool_owner_backup.slot
```

BUT do not see problems with `hardfork_epoch` . We will deal with this in separate PR to my best knowledge

**Comment by paweljakubas** [comment]
I will need to wait for mainnet test when https://github.com/input-output-hk/cardano-wallet/pull/1943 is merged - to my best knowledge

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Network.hs [hunk 7]
```diff
@@ -46,4 +48,5 @@ spec = do
             [ expectField (#decentralizationLevel) (`shouldBe` d)
             , expectField (#desiredPoolNumber) (`shouldBe` nOpt)
             , expectField (#minimumUtxoValue) (`shouldBe` minUtxoValue)
+            , expectField (#hardforkEpochInfo) (`shouldNotBe` Nothing)
```
Comment (KtorZ) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Network.hs]: Will this one be ever true? Since we only run the test scenarios AFTER the hard-fork has happened, I'd assume that `Nothing` will be returned forever. No ?

File: lib/core/src/Cardano/Wallet/DB/Sqlite/Types.hs [hunk 8]
```diff
@@ -302,6 +297,32 @@ instance PathPiece SlotNo where
...
+deriving via Word32 instance (PersistFieldSql EpochNo)
+
+mkEpochNo :: Word32 -> Either Text EpochNo
+mkEpochNo n
+    | isValidEpochNo c = Right c
+    | otherwise = Left . T.pack $ "not a valid epoch number: " <> show n
+    where c = unsafeEpochNo n
+
+persistEpochNo :: EpochNo -> PersistValue
+persistEpochNo = toPersistValue . fromIntegral @Word31 @Word32 . unEpochNo
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/DB/Sqlite/Types.hs:312]: Looks better :smile:

File: specifications/api/swagger.yaml [hunk 9]
```diff
@@ -848,6 +848,7 @@ components:
         decentralization_level: *percentage
         desired_pool_number: *stakePoolsNumber
         minimum_utxo_value: *amount
+        hardfork_epoch_info: *epochInfo
```
Comment (KtorZ) [inline specifications/api/swagger.yaml]: What about just `hardfork` ? or `hardfork_at` for consistency across the API namings?

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Network.hs (hunk 7 above)
Comment (paweljakubas) [reply lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Network.hs]: it was Nothing...if you run this test once! ie., one has to wait for tip client waiting to make local query
```
stack test --ta '-m "NETWORK_PARAMS - Able to fetch network parameters"' cardano-wallet-shelley:integration 
```
But when running several tests then I have Just ... so I used eventually

File: specifications/api/swagger.yaml (hunk 9 above)
Comment (paweljakubas) [reply specifications/api/swagger.yaml]: ok, will change very soon

Comment (paweljakubas) [reply specifications/api/swagger.yaml]: done in https://github.com/input-output-hk/cardano-wallet/pull/1932/commits/ed5a88542614083270857c69a9cb44e5515ce9b1

**Comment by paweljakubas** [comment]
setup: cardano-wallet-shelley produced with this branch and connected to the mainnet
before 18:30 UTC : 
```
cardano-wallet-shelley network parameters
no  "hardfork_at" in response
```
after 18:30 UTC:
```
     "hardfork_at": {
        "epoch_start_time": "2020-07-29T21:44:51Z",
        "epoch_number": 208
    },
```
Thanks @Anviking  and @piotr-iohk  with helping in verifying this!

**Review by KtorZ — APPROVED** [review]

**Comment by KtorZ** [comment]
Unit tests running OOM. Reported in the spreadsheet, hopefully fixed by #1954.
